File: backend/src/main/java/pl/babastudiobe/event/EventService.java
package pl.babastudiobe.event;

import java.util.List;
import java.util.UUID;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.server.ResponseStatusException;

import pl.babastudiobe.media.MediaAsset;
import pl.babastudiobe.media.MediaStorageService;

import static org.springframework.http.HttpStatus.NOT_FOUND;

@Service
class EventService {

	private static final String EVENT_MEDIA_CATEGORY = "events";
	private static final String HOST_MEDIA_CATEGORY = "event-hosts";

	private final EventRepository repository;
	private final MediaStorageService mediaStorageService;

	EventService(EventRepository repository, MediaStorageService mediaStorageService) {
		this.repository = repository;
		this.mediaStorageService = mediaStorageService;
	}

	@Transactional(readOnly = true)
	List<EventResponse> list() {
		return repository.findAllByOrderByEventStartAtAscCreatedAtAsc().stream()
				.map(EventResponse::card)
				.toList();
	}

	@Transactional(readOnly = true)
	EventResponse get(UUID id) {
		return repository.findById(id)
				.map(EventResponse::detail)
				.orElseThrow(() -> new ResponseStatusException(NOT_FOUND, "Nie znaleziono wydarzenia."));
	}

	@Transactional
	EventResponse create(EventDetails details, MultipartFile image, MultipartFile hostImage) {
		EventDetails normalized = details.normalized();
		Event event = new Event(
				normalized.title(),
				normalized.hostName(),
				normalized.shortDescription(),
				normalized.description(),
				normalized.hostDescription(),
				normalized.eventStartAt(),
				normalized.durationMinutes(),
				normalized.location(),
				normalized.capacity(),
				normalized.price(),
				normalized.signupUrl(),
				normalized.titleEn(),
				normalized.shortDescriptionEn(),
				normalized.descriptionEn(),
				normalized.hostDescriptionEn(),
				normalized.locationEn(),
				normalized.priceEn(),
				mediaStorageService.storeImage(EVENT_MEDIA_CATEGORY, image),
				storeIfPresent(HOST_MEDIA_CATEGORY, hostImage)
		);

		return EventResponse.detail(repository.save(event));
	}

	/**
	 * Zdjęcie wydarzenia jest obowiązkowe, więc można je tylko podmienić. Zdjęcie
	 * prowadzącego jest opcjonalne i dlatego ma dodatkowo flagę czyszczącą - bez niej
	 * omyłkowo dodane zdjęcie zostałoby w bazie na zawsze.
	 */
	@Transactional
	EventResponse update(
			UUID id,
			EventDetails details,
			boolean removeHostImage,
			MultipartFile image,
			MultipartFile hostImage
	) {
		Event event = findOrThrow(id);
		EventDetails normalized = details.normalized();

		event.updateDetails(
				normalized.title(),
				normalized.hostName(),
				normalized.shortDescription(),
				normalized.description(),
				normalized.hostDescription(),
				normalized.eventStartAt(),
				normalized.durationMinutes(),
				normalized.location(),
				normalized.capacity(),
				normalized.price(),
				normalized.signupUrl(),
				normalized.titleEn(),
				normalized.shortDescriptionEn(),
				normalized.descriptionEn(),
				normalized.hostDescriptionEn(),
				normalized.locationEn(),
				normalized.priceEn()
		);

		MediaAsset newImage = storeIfPresent(EVENT_MEDIA_CATEGORY, image);
		if (newImage != null) {
			mediaStorageService.delete(event.replaceImage(newImage));
		}

		MediaAsset newHostImage = storeIfPresent(HOST_MEDIA_CATEGORY, hostImage);
		if (newHostImage != null) {
			mediaStorageService.delete(event.replaceHostImage(newHostImage));
		}
		else if (removeHostImage) {
			mediaStorageService.delete(event.replaceHostImage(null));
		}

		return EventResponse.detail(repository.save(event));
	}

	@Transactional
	void delete(UUID id) {
		Event event = findOrThrow(id);
		MediaAsset image = event.getImage();
		MediaAsset hostImage = event.getHostImage();

		repository.delete(event);
		repository.flush();

		mediaStorageService.delete(image);
		mediaStorageService.delete(hostImage);
	}

	private Event findOrThrow(UUID id) {
		return repository.findById(id)
				.orElseThrow(() -> new ResponseStatusException(NOT_FOUND, "Nie znaleziono wydarzenia."));
	}

	private MediaAsset storeIfPresent(String category, MultipartFile file) {
		return file == null || file.isEmpty() ? null : mediaStorageService.storeImage(category, file);
	}
}

File: frontend/src/app/core/event.service.ts
import { HttpClient } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';

import { API_BASE_URL } from './api-url';
import { CreateStudioEventInput, StudioEvent, UpdateStudioEventInput } from './studio-event';

@Injectable({
  providedIn: 'root'
})
export class EventService {
  private readonly apiUrl: string;

  constructor(
    private readonly http: HttpClient,
    @Inject(API_BASE_URL) apiBaseUrl: string
  ) {
    this.apiUrl = `${apiBaseUrl}/api`;
  }

  list() {
    return this.http.get<StudioEvent[]>(`${this.apiUrl}/events`);
  }

  get(id: string) {
    return this.http.get<StudioEvent>(`${this.apiUrl}/events/${id}`);
  }

  create(input: CreateStudioEventInput) {
    const formData = this.toFormData(input);
    formData.append('image', input.image);

    return this.http.post<StudioEvent>(`${this.apiUrl}/admin/events`, formData);
  }

  update(id: string, input: UpdateStudioEventInput) {
    const formData = this.toFormData(input);

    // Brak pola pliku oznacza dla backendu "zostaw dotychczasowe zdjęcie".
    if (input.image) {
      formData.append('image', input.image);
    }

    if (input.removeHostImage) {
      formData.append('removeHostImage', 'true');
    }

    return this.http.put<StudioEvent>(`${this.apiUrl}/admin/events/${id}`, formData);
  }

  delete(id: string) {
    return this.http.delete<void>(`${this.apiUrl}/admin/events/${id}`);
  }

  private toFormData(input: UpdateStudioEventInput) {
    const formData = new FormData();
    formData.append('title', input.title);
    formData.append('hostName', input.hostName);
    formData.append('shortDescription', input.shortDescription);
    formData.append('description', input.description);
    formData.append('eventStartAt', input.eventStartAt);

    this.appendOptional(formData, 'hostDescription', input.hostDescription);
    this.appendOptional(formData, 'durationMinutes', input.durationMinutes?.toString() ?? '');
    this.appendOptional(formData, 'location', input.location);
    this.appendOptional(formData, 'capacity', input.capacity?.toString() ?? '');
    this.appendOptional(formData, 'price', input.price);
    this.appendOptional(formData, 'signupUrl', input.signupUrl);

    // Puste tłumaczenie wysyłamy jako pusty ciąg - backend zamienia je na null.
    formData.append('titleEn', input.titleEn ?? '');
    formData.append('shortDescriptionEn', input.shortDescriptionEn ?? '');
    formData.append('descriptionEn', input.descriptionEn ?? '');
    formData.append('hostDescriptionEn', input.hostDescriptionEn ?? '');
    formData.append('locationEn', input.locationEn ?? '');
    formData.append('priceEn', input.priceEn ?? '');

    if (input.hostImage) {
      formData.append('hostImage', input.hostImage);
    }

    return formData;
  }

  private appendOptional(formData: FormData, key: string, value: string) {
    const trimmedValue = value.trim();
    if (trimmedValue) {
      formData.append(key, trimmedValue);
    }
  }
}
